File: solved/RemoveNthFromEnd.go
// Package main has solved algorithms
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/93/linked-list/603/
package main

// import log "github.com/sirupsen/logrus"

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	find := head
	node := head
	prev := node

	for n > 1 {
		find = find.Next
		n--
	}

	for find.Next != nil {
		prev = node
		node = node.Next
		find = find.Next
	}

	if node == head {
		return head.Next
	}

	prev.Next = node.Next

	return head
}

func main() {
	// log.Info(removeNthFromEnd())
}

File: solved/typescript/removeNthFromEnd/removeNthFromEnd.ts
import { ListNode } from '../lib/types/ListNode'

// https://leetcode.com/explore/learn/card/linked-list/214/two-pointer-technique/1296/
type MaybeListNode = ListNode | null
const removeNthFromEnd = (head: MaybeListNode, n: number): MaybeListNode => {
  if (!head) return null
  let slow: MaybeListNode = head
  let fast: MaybeListNode = head
  while (n-- >= 0) {
    if (!fast) return head.next
    fast = fast?.next || null
  }

  while (fast) {
    fast = fast?.next || null
    slow = slow!.next
  }

  slow!.next = slow!.next!.next
  return head
}

export { removeNthFromEnd }
